class.c

```diff
@@ -1782,26 +1778,25 @@ singleton_class_of(VALUE obj)
 {
     VALUE klass;
 
-    if (FIXNUM_P(obj) || FLONUM_P(obj) || STATIC_SYM_P(obj)) {
-      no_singleton:
+    switch (TYPE(obj)) {
```

Comment:
I think the previous code generated better code than `TYPE()`.

Reply:
@nobu tell us how better?  I don't see any clear problem

```diff
--- miniruby.master 2020-06-22 14:11:30.015038077 +0900
+++ miniruby        2020-06-22 14:11:18.846865697 +0900
@@ -1,102 +1,116 @@
 
-miniruby.master:	file format ELF64-x86-64
+miniruby:	file format ELF64-x86-64
 
 
 Disassembly of section .text:
 
 singleton_class_of:
+ 55                                    	pushq	%rbp
  41 57                                 	pushq	%r15
  41 56                                 	pushq	%r14
  53                                    	pushq	%rbx
+ 50                                    	pushq	%rax
  49 89 fe                              	movq	%rdi, %r14
- 41 80 fe 0c                           	cmpb	$12, %r14b
- 0f 84 0c 01 00 00                     	je	268 <singleton_class_of+0x11e>
- 44 89 f0                              	movl	%r14d, %eax
- 83 e0 01                              	andl	$1, %eax
- 48 85 c0                              	testq	%rax, %rax
- 0f 85 fd 00 00 00                     	jne	253 <singleton_class_of+0x11e>
- 44 89 f0                              	movl	%r14d, %eax
- 83 e0 03                              	andl	$3, %eax
- 83 f8 02                              	cmpl	$2, %eax
- 0f 84 ee 00 00 00                     	je	238 <singleton_class_of+0x11e>
  41 f6 c6 07                           	testb	$7, %r14b
- 0f 85 a8 00 00 00                     	jne	168 <singleton_class_of+0xe2>
- 4c 89 f0                              	movq	%r14, %rax
- 48 83 e0 f7                           	andq	$-9, %rax
- 0f 84 9b 00 00 00                     	je	155 <singleton_class_of+0xe2>
+ 0f 95 c0                              	setne	%al
+ 48 f7 c7 f7 ff ff ff                  	testq	$-9, %rdi
+ 40 0f 94 c5                           	sete	%bpl
+ 40 08 c5                              	orb	%al, %bpl
+ 40 80 fd 01                           	cmpb	$1, %bpl
+ 75 21                                 	jne	33 <singleton_class_of+0x46>
+ 49 83 fe 13                           	cmpq	$19, %r14
+ 7f 3e                                 	jg	62 <singleton_class_of+0x69>
+ 4d 85 f6                              	testq	%r14, %r14
+ 75 66                                 	jne	102 <singleton_class_of+0x96>
+ 48 8d 05 19 1c 54 00                  	leaq	5512217(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 04 01 00 00                     	jne	260 <singleton_class_of+0x148>
+ eb 6a                                 	jmp	106 <singleton_class_of+0xb0>
  49 8b 06                              	movq	(%r14), %rax
  89 c1                                 	movl	%eax, %ecx
  83 e1 1f                              	andl	$31, %ecx
- 83 c1 fc                              	addl	$-4, %ecx
- 83 f9 10                              	cmpl	$16, %ecx
- 77 1b                                 	ja	27 <singleton_class_of+0x72>
- 48 8d 15 02 86 24 00                  	leaq	2393602(%rip), %rdx
- 48 63 0c 8a                           	movslq	(%rdx,%rcx,4), %rcx
- 48 01 d1                              	addq	%rdx, %rcx
- ff e1                                 	jmpq	*%rcx
+ 83 f9 15                              	cmpl	$21, %ecx
+ 0f 87 83 00 00 00                     	ja	131 <singleton_class_of+0xda>
+ ba 00 00 0e 00                        	movl	$917504, %edx
+ 0f a3 ca                              	btl	%ecx, %edx
+ 73 24                                 	jae	36 <singleton_class_of+0x85>
+ 49 83 fe 14                           	cmpq	$20, %r14
+ 74 08                                 	je	8 <singleton_class_of+0x6f>
+ eb 47                                 	jmp	71 <singleton_class_of+0xb0>
+ 49 83 fe 14                           	cmpq	$20, %r14
+ 75 52                                 	jne	82 <singleton_class_of+0xc1>
+ 48 8d 05 ca 1b 54 00                  	leaq	5512138(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 c5 00 00 00                     	jne	197 <singleton_class_of+0x148>
+ eb 2b                                 	jmp	43 <singleton_class_of+0xb0>
+ 83 f9 05                              	cmpl	$5, %ecx
+ 75 46                                 	jne	70 <singleton_class_of+0xd0>
  a9 00 00 00 20                        	testl	$536870912, %eax
- 0f 85 ac 00 00 00                     	jne	172 <singleton_class_of+0x11e>
+ 74 49                                 	je	73 <singleton_class_of+0xda>
+ e9 c0 00 00 00                        	jmp	192 <singleton_class_of+0x156>
+ 49 83 fe 08                           	cmpq	$8, %r14
+ 75 2b                                 	jne	43 <singleton_class_of+0xc7>
+ 48 8d 05 85 1b 54 00                  	leaq	5512069(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 98 00 00 00                     	jne	152 <singleton_class_of+0x148>
+ 48 8d 3d e0 68 24 00                  	leaq	2386144(%rip), %rdi
+ 4c 89 f6                              	movq	%r14, %rsi
+ 31 c0                                 	xorl	%eax, %eax
+ e8 24 47 06 00                        	callq	411428 <rb_bug>
+ 49 83 fe 34                           	cmpq	$52, %r14
+ 74 13                                 	je	19 <singleton_class_of+0xda>
+ 41 f6 c6 01                           	testb	$1, %r14b
+ e9 86 00 00 00                        	jmp	134 <singleton_class_of+0x156>
+ b8 10 04 30 00                        	movl	$3146768, %eax
+ 0f a3 c8                              	btl	%ecx, %eax
+ 72 7c                                 	jb	124 <singleton_class_of+0x156>
  49 8b 5e 08                           	movq	8(%r14), %rbx
  f6 c3 07                              	testb	$7, %bl
- 75 32                                 	jne	50 <singleton_class_of+0xad>
- 49 89 df                              	movq	%rbx, %r15
- 49 83 e7 f7                           	andq	$-9, %r15
- 74 29                                 	je	41 <singleton_class_of+0xad>
+ 75 32                                 	jne	50 <singleton_class_of+0x115>
+ 48 89 d8                              	movq	%rbx, %rax
+ 48 83 e0 f7                           	andq	$-9, %rax
+ 74 29                                 	je	41 <singleton_class_of+0x115>
  48 8b 03                              	movq	(%rbx), %rax
  89 c1                                 	movl	%eax, %ecx
  83 e1 1f                              	andl	$31, %ecx
  83 f9 1b                              	cmpl	$27, %ecx
- 74 1c                                 	je	28 <singleton_class_of+0xad>
+ 74 1c                                 	je	28 <singleton_class_of+0x115>
  25 00 10 00 00                        	andl	$4096, %eax
  48 85 c0                              	testq	%rax, %rax
- 74 12                                 	je	18 <singleton_class_of+0xad>
+ 74 12                                 	je	18 <singleton_class_of+0x115>
  be 01 0c 00 00                        	movl	$3073, %esi
  48 89 df                              	movq	%rbx, %rdi
- e8 98 e1 1f 00                        	callq	2089368 <rb_ivar_get>
+ e8 e0 c1 1f 00                        	callq	2081248 <rb_ivar_get>
  4c 39 f0                              	cmpq	%r14, %rax
- 74 1a                                 	je	26 <singleton_class_of+0xc7>
+ 74 13                                 	je	19 <singleton_class_of+0x128>
  4c 8b 7b 20                           	movq	32(%rbx), %r15
  4c 89 f7                              	movq	%r14, %rdi
- e8 27 dc ff ff                        	callq	-9177 <rb_make_metaclass>
+ e8 3f db ff ff                        	callq	-9409 <rb_make_metaclass>
  48 89 c3                              	movq	%rax, %rbx
  4c 89 78 20                           	movq	%r15, 32(%rax)
- 49 89 c7                              	movq	%rax, %r15
- 49 83 e7 f7                           	andq	$-9, %r15
  49 8b 06                              	movq	(%r14), %rax
+ 40 84 ed                              	testb	%bpl, %bpl
  25 00 08 00 00                        	andl	$2048, %eax
  f6 c3 07                              	testb	$7, %bl
- 75 03                                 	jne	3 <singleton_class_of+0xd7>
- 4d 85 ff                              	testq	%r15, %r15
+ 75 07                                 	jne	7 <singleton_class_of+0x13f>
+ 48 89 d9                              	movq	%rbx, %rcx
+ 48 83 e1 f7                           	andq	$-9, %rcx
  48 8b 0b                              	movq	(%rbx), %rcx
  48 09 c1                              	orq	%rax, %rcx
  48 89 0b                              	movq	%rcx, (%rbx)
- eb 33                                 	jmp	51 <singleton_class_of+0x115>
- 49 83 fe 14                           	cmpq	$20, %r14
- 74 1d                                 	je	29 <singleton_class_of+0x105>
- 49 83 fe 08                           	cmpq	$8, %r14
- 74 0e                                 	je	14 <singleton_class_of+0xfc>
- 4d 85 f6                              	testq	%r14, %r14
- 75 43                                 	jne	67 <singleton_class_of+0x136>
- 48 8d 05 be 3f 54 00                  	leaq	5521342(%rip), %rax
- eb 10                                 	jmp	16 <singleton_class_of+0x10c>
- 48 8d 05 8d 3f 54 00                  	leaq	5521293(%rip), %rax
- eb 07                                 	jmp	7 <singleton_class_of+0x10c>
- 48 8d 05 9c 3f 54 00                  	leaq	5521308(%rip), %rax
- 48 8b 18                              	movq	(%rax), %rbx
- 48 83 fb 08                           	cmpq	$8, %rbx
- 74 21                                 	je	33 <singleton_class_of+0x136>
  48 89 d8                              	movq	%rbx, %rax
+ 48 83 c4 08                           	addq	$8, %rsp
  5b                                    	popq	%rbx
  41 5e                                 	popq	%r14
  41 5f                                 	popq	%r15
+ 5d                                    	popq	%rbp
  c3                                    	retq
- 48 8d 05 5b 19 54 00                  	leaq	5511515(%rip), %rax
+ 48 8d 05 d3 f4 53 00                  	leaq	5502163(%rip), %rax
  48 8b 38                              	movq	(%rax), %rdi
- 48 8d 35 49 88 24 00                  	leaq	2394185(%rip), %rsi
- 31 c0                                 	xorl	%eax, %eax
- e8 9a 5b 06 00                        	callq	416666 <rb_raise>
- 48 8d 3d 52 88 24 00                  	leaq	2394194(%rip), %rdi
- 4c 89 f6                              	movq	%r14, %rsi
+ 48 8d 35 19 68 24 00                  	leaq	2385945(%rip), %rsi
  31 c0                                 	xorl	%eax, %eax
- e8 9e 66 06 00                        	callq	419486 <rb_bug>
- 66 0f 1f 84 00 00 00 00 00            	nopw	(%rax,%rax)
+ e8 62 3b 06 00                        	callq	408418 <rb_raise>
+ 66 90                                 	nop
```

Reply:
That was diff of `singleton_class_of`, both compiled using `clang-11 -O3`.

Reply:
I see, the change of `rb_type`, clang, or both, did the thing.